File: packages/node/src/routing/handlers/type.ts
import type { Update } from "../../kernel/update.js";
import { CallbackContext } from "../../kernel/context.js";
import type { RawUpdate } from "../../client/types.js";
import { BaseHandler, HandlerCallback } from "./base.js";

/**
 * Handler that matches updates based on an arbitrary type predicate function.
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 * @example
 * ```ts
 * const customTypeHandler = new TypeHandler(
 *   (update) => Boolean(update.message_reaction),
 *   async (update, context) => { console.log("Reaction update received!"); }
 * );
 * ```
 */
export class TypeHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /**
   * The custom type predicate function.
   */
  public readonly typePredicate: (update: Update | RawUpdate) => boolean | Promise<boolean>;

  /**
   * Constructs a new {@link TypeHandler}.
   *
   * @param typePredicate - Function that evaluates whether an update matches.
   * @param callback - Function invoked on match.
   */
  constructor(
    typePredicate: (update: Update | RawUpdate) => boolean | Promise<boolean>,
    callback: HandlerCallback<C, R>,
  ) {
    super(callback);
    this.typePredicate = typePredicate;
  }

  /**
   * Evaluates the type predicate against the incoming update.
   *
   * @param update - Incoming Telegram update.
   * @returns `true` if the type predicate is satisfied.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    return Boolean(await this.typePredicate(update));
  }
}

File: packages/python/src/telebot_py/routing/handlers/type.py
"""Handler matching updates via an arbitrary type predicate."""

from __future__ import annotations

from collections.abc import Callable

from telebot_py.routing.handlers.base import BaseHandler, HandlerCallback
from telebot_py.types import Update

#: Predicate deciding whether an update matches a :class:`TypeHandler`.
TypePredicate = Callable[[Update], bool]


class TypeHandler(BaseHandler):
    """Handler that matches updates based on an arbitrary type predicate.

    Any-update pass-through: the predicate receives the full update and its
    boolean result decides the match, mirroring the node ``TypeHandler``.

    Attributes:
        type_predicate: Function evaluating whether an update matches.
    """

    def __init__(
        self,
        type_predicate: TypePredicate,
        callback: HandlerCallback,
        *,
        block: bool = True,
    ) -> None:
        """Initialize the handler.

        Args:
            type_predicate: Function that evaluates whether an update
                matches.
            callback: Callable invoked when the predicate passes.
            block: Whether dispatch awaits the callback before moving on.
        """
        super().__init__(callback, block=block)
        self.type_predicate = type_predicate

    def check_update(self, update: object) -> bool:
        """Return whether the type predicate is satisfied by the update.

        Args:
            update: The incoming update.

        Returns:
            ``True`` when the predicate returns a truthy value for the
            update, ``False`` otherwise (including for non-``Update``
            objects).
        """
        if not isinstance(update, Update):
            return False
        return bool(self.type_predicate(update))
